src/ripple/server/impl/io_list.h

```diff
@@ -0,0 +1,272 @@
+                closed. No finisher is invoked in this case.
+
+            Blocks until all work is destroyed.
+    */
+    ~io_list()
+    {
+        destroy();
+    }
+
+    /** Return `true` if the list is closed.
+
+        Thread Safety:
+            Undefined result if called concurrently
+            with close().
+    */
+    bool
+    closed() const
+    {
+        return closed_;
+    }
+
+    /** Create associated work if not closed.
+
+        Requirements:
+            `std::is_base_of_v<work, T> == true`
+
+        Thread Safety:
+            May be called concurrently.
+
+        Effects:
+            Atomically creates, inserts, and returns new
+            work T, or returns nullptr if the io_list is
+            closed,
+
+        If the call succeeds and returns a new object,
+        it is guaranteed that a subsequent call to close
+        will invoke work::close on the object.
+
+    */
+    template <class T, class... Args>
+    std::shared_ptr<T>
+    emplace(Args&&... args);
+
+    /** Cancel active I/O.
+
+        Thread Safety:
+            May not be called concurrently.
+
+        Effects:
+            Associated work is closed.
+
+            Finisher if provided, will be called when
+            all associated work is destroyed. The finisher
+            may be called from a foreign thread, or within
+            the call to this function.
+
+            Only the first call to close will set the
+            finisher.
+
+            No effect after the first call.
+    */
+    template<class Finisher>
+    void
+    close(Finisher&& f);
+
+    void
+    close()
+    {
+        close([]{});
+    }
+
+    /** Block until the io_list stops.
+        
+        Effects:
+            The caller is blocked until the io_list is
+            closed and all associated work is destroyed.
+
+        Thread safety:
+            May be called concurrently.
+
+        Preconditions:
+            No call to io_service::run on any io_service
+            used by work objects associated with this io_list
+            exists in the caller's call stack.
+    */
+    template<class = void>
+    void
+    join();
+};
+
+//------------------------------------------------------------------------------
+
+template<class>
+void
+io_list::work::destroy()
+{
+    if(! ios_)
+        return;
+    std::function<void(void)> f;
+    {
+        std::lock_guard<
+            std::mutex> lock(ios_->m_);
+        ios_->map_.erase(this);
+        if(--ios_->n_ == 0 &&
+            ios_->closed_)
+        {
+            std::swap(f, ios_->f_);
+            ios_->cv_.notify_all();
+        }
+    }
+    if(f)
+        f();
+}
+
+template<class>
+void
+io_list::destroy()
+{
+    close();
+    join();
+}
+
+template <class T, class... Args>
+std::shared_ptr<T>
+io_list::emplace(Args&&... args)
+{
+    static_assert(std::is_base_of<work, T>::value,
+        "T must derive from io_list::work");
+    if(closed_)
+        return nullptr;
```

Comment:
We're reading `closed_` without a lock. (we discussed this earlier, so you know the issues - I'm OK if you want to keep this as is).